http: new body transform filter with the substitution formatter

wbpcode opened the pull request:
Commit Message: http: new body transform filter with the substitution formatter
Additional Description:

To close https://github.com/envoyproxy/envoy/issues/35783. There is enough context in the #35783. In short sentence, this PR add a transform filter to modify request or response headers by the substitution formatter API.

It's also possible to add body attributes to headers and refresh the routes.

Risk Level: low. New HTTP filter.
Testing: unit/integration.
Docs Changes: added.
Release Notes: added.
Platform Specific Features: n/a.

wbpcode commented:
cc @arkodg cc @agrawroh for you guys may care about this. cc @botengyao for this may helpful for AI traffic routing.

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+// Configuration for the transform filter. The request and response transformations are
+// independent and could be configured separately.
+// Only JSON body transformation is supported for now.
+message TransformConfig {
```
nit: move this to be the first message as this is the filter's config.

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+  // the current request and force re-evaluation of the route. This is more expensive and
+  // should only be used when the route match criteria depends on the transformed headers.
+  //
+  // One one of ``clear_cluster_cache`` and ``clear_route_cache`` can be true.
```
```suggestion
  // Only one of ``clear_cluster_cache`` and ``clear_route_cache`` can be true.
```

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+  bool clear_cluster_cache = 3;
+
+  // If true and the request headers are transformed, Envoy will clear the route cache for
+  // the current request and force re-evaluation of the route. This is more expensive and
```
```suggestion
  // the current request and force re-evaluation of the route. This has performance penalty and
```

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+
+  // If true and the request headers are transformed, Envoy will re-evaluate the target
+  // cluster in the same route. Please ensure the cluster specifier in the route supports
+  // dynamic evaluation, e.g., ``envoy.router.cluster_specifier_plugin.matcher``.
```
What would happen if there isn't a cluster-specifier-plugin-matcher defined?
Also, please add a reference to the matcher specification.

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+  // * ``%RQ_BODY(X)%``: the request body. And ``X`` is the JSON pointer to extract a specific
+  //   field from the JSON body. If ``X`` is empty, the whole body will be used.
+  // * ``%RS_BODY(X)%``: the response body. And ``X`` is the JSON pointer to extract a specific
+  //   field from the JSON body. If ``X`` is empty, the whole body will be used.
```
> the whole body will be used

FWIW this might cause the headermap to go over the limit, and may be an attack vector, but as this filter doesn't have a security posture, it might be ok.

Also it would be good to clarify whether using this filter implies that all requests/responses are buffered, even if headers_mutations are defined without referencing RQ/RS_BODY. If there's a need to support both buffering and streaming modes, consider adding an enum field specifying the mode (and maybe only support one for now).

adisuissa reviewed (commented):
Thanks. Left a few API comments

arkodg commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+
+  // Body transformation configuration. If not set, no body transformation will be performed.
+  // NOTE the ``RQ_BODY`` and ``RS_BODY`` format specifiers can also be used here.
+  config.core.v3.SubstitutionFormatString body_format_string = 3;
```
thoughts on using JSONPatch semantics here ?
https://jsonpatch.com
its used in Envoy Gateway API to edit xDS https://gateway.envoyproxy.io/docs/tasks/extensibility/envoy-patch-policy/#customize-response

wbpcode commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,69 @@
+  // NOTE the ``RQ_BODY`` and ``RS_BODY`` format specifiers can also be used here.
+  config.core.v3.SubstitutionFormatString body_format_string = 3;
+
+  // If set the result of ``body_format_string`` will be treated as a patch and will be merged
```
SGTM. This is a great suggestion.

botengyao commented on source/extensions/filters/http/transform/transform.cc:
```diff
@@ -0,0 +1,387 @@
+    decoder_callbacks_->addDecodedData(data, true);
+    handleCompleteRequestBody();
+    return Http::FilterDataStatus::Continue;
+  }
```
Related but not tied / blocker to this PR, do you have some insights for the streaming JSON parser support without waiting for the whole body once we get the enough data. I think the current nlohmann lib doesn't quite support it, and we could need some SAX style approach.

wbpcode commented:
friendly ping for a new review cc @adisuissa

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+// Only JSON body transformation is supported for now.
+message TransformConfig {
+  // Configuration for transforming request. If set then the request will be buffered
+  // until the entire body is received and transformed for JSON request bodies.
```
Do you think this needs to also address HTTP trailers case?
(I'm not saying to add it now, but there may need to be a config knob in the future that decides what needs to be buffered before sending it down/up the stream. I believe this may be in the "Transformation" message).

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+// HTTP header and body transformations.
+// [#extension: envoy.filters.http.transform]
+
+// Configuration for the transform filter. The request and response transformations are
```
Consider adding some clarification that this filter will buffer the request/response until the entire body is received:

```suggestion
// Configuration for the transform filter. The filter may buffer the request/response until the entire body is received, and then mutate the headers and body according to the contents of the request/response. The request and response transformations are
```

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+  //   lookup key in the namespace with the option of specifying nested keys separated by ':'.
+  // * ``%RESPONSE_BODY(KEY*)%``: the response body. And ``Key`` KEY is an optional
+  //   lookup key in the namespace with the option of specifying nested keys separated by ':'.
+  repeated config.common.mutation_rules.v3.HeaderMutation headers_mutations = 1;
```
Will the headers be blocked until the body is buffered even if `headers_mutations` isn't configured?

adisuissa replied:
I understand that this is an implementation issue, but I think it should be spelled clearly in a note above (example: buffering of the entire request headers and body will always happen on a request even if only headers are transformed).

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+
+  // Body transformation configuration. If not set, no body transformation will be performed.
+  // NOTE the ``REQUEST_BODY`` and ``RESPONSE_BODY`` format specifiers can also be used here.
+  config.core.v3.SubstitutionFormatString body_format = 3
```
```suggestion
  config.core.v3.SubstitutionFormatString body_format = 1
```

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+      [(validate.rules).message = {required: true}];
+
+  // The action to perform for body transformation. Default is ``MERGE``.
+  TransformAction action = 4;
```
```suggestion
  TransformAction action = 2;
```

wbpcode commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+
+  // Body transformation configuration. If not set, no body transformation will be performed.
+  // NOTE the ``REQUEST_BODY`` and ``RESPONSE_BODY`` format specifiers can also be used here.
+  config.core.v3.SubstitutionFormatString body_format = 3
```
The `Substitution` in  `SubstitutionFormatString` is means the substitution formatter will be used in the format string. For example, in the final generated output, `%REQ(xxx)%` in the format string will be substituted by the actual header value. 

And the `action` is used to indicate how to handle the final generated output of the `body_format` and the original body content.

I will add some more comment to make it more clear.

adisuissa commented on api/envoy/extensions/filters/http/transform/v3/transform.proto:
```diff
@@ -0,0 +1,83 @@
+  repeated config.common.mutation_rules.v3.HeaderMutation headers_mutations = 1;
+
+  // The body transformation configuration. If not set, no body transformation will be performed.
+  BodyTransformation body_transformation = 2;
```
I guess the same can be said about headers, but I agree that headers are more likely to have multiple transformations (one for each header key).

adisuissa reviewed (commented):
Overall API LGTM.
Left a suggestion to clarify the implication of using this filter (buffering) in a note (for Envoy at least).

wbpcode commented:
We added a new type alias in this PR's previous commit. But another alias for same type was added in the main by recently merged PR. So, after merged the main, at the latest commit, we removed the unnecessary new alias and reuse the another alias.
